server/blob/blob-union/src/main/java/org/apache/james/blob/union/HybridBlobStore.java

```diff
@@ -128,45 +145,62 @@ private HybridBlobStore(BlobStore lowCostBlobStore, BlobStore highPerformanceBlo
 
     @Override
     public Mono<BlobId> save(BucketName bucketName, byte[] data, StoragePolicy storagePolicy) {
-        return selectBlobStore(storagePolicy, Mono.just(data.length > configuration.getSizeThreshold()))
-            .flatMap(blobStore -> Mono.from(blobStore.save(bucketName, data, storagePolicy)));
+        return save(bucketName, new ByteArrayInputStream(data), storagePolicy);
     }
 
     @Override
     public Mono<BlobId> save(BucketName bucketName, InputStream data, StoragePolicy storagePolicy) {
         Preconditions.checkNotNull(data);
 
-        BufferedInputStream bufferedInputStream = new BufferedInputStream(data, configuration.getSizeThreshold() + 1);
-        return selectBlobStore(storagePolicy, Mono.fromCallable(() -> isItABigStream(bufferedInputStream)))
-            .flatMap(blobStore -> Mono.from(blobStore.save(bucketName, bufferedInputStream, storagePolicy)));
-    }
-
-    private Mono<BlobStore> selectBlobStore(StoragePolicy storagePolicy, Mono<Boolean> largeData) {
         switch (storagePolicy) {
             case LOW_COST:
-                return Mono.just(lowCostBlobStore);
+                return Mono.from(lowCostBlobStore.save(bucketName, data, storagePolicy))
+                    .doFinally(any -> closeInputStream(data));
+
             case SIZE_BASED:
-                return largeData.map(isLarge -> {
-                    if (isLarge) {
-                        return lowCostBlobStore;
-                    }
-                    return highPerformanceBlobStore;
-                });
-            case HIGH_PERFORMANCE:
-                return Mono.just(highPerformanceBlobStore);
+                BufferedInputStream bufferedInputStream = new BufferedInputStream(data, configuration.getSizeThreshold() + 1);
+                return Mono.fromCallable(() -> isLargeStream(bufferedInputStream))
+                    .filter(Boolean::booleanValue)
```

Comment:
Let me suggest :

```
Mono.just(bufferedInputStream)
    .filterWhen(this::isLargeStream) // isLargeStream thus needs to return a mono and should not throw
    .flatMap(...)
```

Reply:
If we do that way, we also moving from `reactive condition` to `normal condition` because `bufferedInputStream` used in both case.
Are you prefer it?